### ISSUE TITLE
question: how to get the last 20% of the genome?

### ISSUE THREAD

**Issue by StefanoLonardi** [marbl/canu#208]
I am trying to assemble a ~620Mb plant genome (essentially haploid, 35% GC) using a 91x PacBio dataset. The first assembly with **CANU** v1.3 (using `corMhapSensitivity=low`, other params are default) gave the following stats:

```
Contig N50 (bp)	 5,307,785 
Contig L50 (bp)	 27 
Contig NG50 (bp)	 3,966,144 
Contig LG50 (bp)	 39 
Total assembled (bp)	 505,857,042 
Number of assembled unitigs	 878 
```

While I am happy with the contiguity, I am a little disappointed by the size of the assembly.
Only 80% of PacBio bases map to the assembly during the polishing step using **BLASR** (Quiver), which would imply that our genome is ~626Mb. I understand that the 20% missing is probably the repetitive portion of the genome, but I would like to know if there anything I can do it to get some of the missing genome.

First, can you please advise on which parameters I should try to change to possibly improve the assembly? Would `corMhapSensitivity=high` help? From the discussions here I gathered that `corMaxEvidenceErate`, `corOutCoverage` and `corMinCoverage` also can affect the quality of the assembly. Would it make any difference in my case?

Second, what if I take the unmapped **Quiver** reads (which are supposed to represent the missing part of my genome) and try to assemble them independently with **CANU**? What params would you use in this case?

Thanks for your help, and for a great assembler! I wrote some of the Celera Assembler code many years ago, so I feel particularly attached to **CANU** :)

**Comment by skoren** [comment]
You can check the asm.unassembled.fasta file for the missing sequence. Ignore anything with 1 read in it since it's a singleton but there will be collapsed repeats with >1 read. I'd expect that to have some of the repeat content that couldn't be assembled which you could include for Quiver mapping. 

I don't think adjusting the sensitivity or other parameters are likely to help because usually the repeats aren't assembled because they're too long for the reads. You could try a run with ```corMhapSensitivity=normal corMaxEvidenceErate=0.15 corOutCoverage=100```. That would only keep the best mappings for the reads and correct more of them so it could potentially separate more repeats and assemble them. As for assembling the unmapped reads, it's worth a shot. You probably won't get much assembly but I'd start with just default parameters and see what that gives you.

**Comment by StefanoLonardi** [comment]
Thank you Sergey. I will give it a try.

**Comment by brianwalenz** [comment]
As for assembling the unassembled reads:

- Use the previously corrected reads (-trim-assemble).
- Increase kmer size, up to 31 (ovlMerSize).
- Increase kmer cutoff threshold (ovlMerThreshold, ovlMerDistinct, ovlMerTotal) to ...?

Generating the kmer histogram for each mersize 22..31 (maybe 22, 25, 28, 31) will help picking both parameters.  You'll want to pick the largest mer size that doesn't have a bunch of unique (count=1) kmers.  For the threshold, a good starting guess would be ovlMerDistinct=0.99 -- ignore the most common 1% of kmers when seeding overlaps.  I (still) haven't written up the method I use for picking this.  The gist is to plot column 3 (fraction distinct) vs 4 (fraction actual) in the meryl histogram output, then finding a knee where fraction actual dominates.

I vaguely remember hearing I got your desk at Celera.

**Comment by StefanoLonardi** [comment]
Thank you, Brian. I will try what you suggest.

**Comment by StefanoLonardi** [comment]
Sergey, the portion of asm.unassembled.fasta with unitigs with >1 read is only 12Mb, while I am missing 120Mb.

**Comment by skoren** [comment]
Ah but that 12mb is likely extremely collapsed, maybe not 10-copies on average but probably at least 2-3. In the extreme cases like rDNA arrays, the assembly has a couple of contigs representing the rDNA when the array is multi-megabasepairs long. I'd check what fraction of the 20% of unmapped data map to this 12Mb.

**Comment by StefanoLonardi** [comment]
Sergey, I will definitely check what fraction of the reads map to this 12Mb.

**Comment by StefanoLonardi** [comment]
Update: I used the RS_Resequencing.1 pipeline (Quiver) in SMRTportal with default params (BLASR max divergence 30%, minimum anchor size 12) to map all the original 91x pacbio reads against (1) the  **CANU** assembly (using `corMhapSensitivity=low`, other params are default) reported at the top of this thread (878 unitigs), and (2) the 12.4Mb unassembled.fasta restricted to unitigs with >1 read (530 unitigs).

Results: The average coverage for (1) is 74.4x, while the average coverage for (2) is 2,775x. I guess this is good news for **CANU**, but I wanted to take a closer look at the coverage unitig by unitig.

I took the two coverage.bed files produced by SMRTportal for (1) and (2) and wrote a script that plots the coverage along the unitigs. I am attaching two PDF files illustrating the coverage for unitigs in (1) and the coverage for unitigs in (2). I sorted the unitigs by average coverage, so I could plot ten on the same page. I have two observations/questions:
A) I was surprised to see assembled unitigs with coverage close to zero
B) While it is clear that unitigs in the assembled portion have (on average) a coverage closer to the expected 70-90x, there are still a lot of unitigs with crazy high coverage (if you scroll at the very end of the PDF document). Should we worry about these contigs being overcollapsed copies?

Comments?

[canu91x_low_default_assembled_unitigs_blasrcoverage.pdf](https://github.com/marbl/canu/files/392695/canu91x_low_default_assembled_unitigs_blasrcoverage.pdf)
[canu91x_low_default_unassembled_unitigs_blasrcoverage.pdf](https://github.com/marbl/canu/files/392696/canu91x_low_default_unassembled_unitigs_blasrcoverage.pdf)

**Comment by skoren** [comment]
We looked into correlation between mis-assemblies detected by optical maps in an assembly with mapped Quiver coverage and didn't find a good correlation. Some mis-assemblies had high coverage and lots of regions had lower mapping coverage but were correct so I think the lower coverage regions in the contigs are OK. If you have some independent evidence like optical mapping or Illumina pair data that indicates coverage drops that would be more informative of a mis-assembly.

**Comment by StefanoLonardi** [comment]
Sergey and Brian: thank you very much for looking into this.

In the meantime, I tried multiple assemblies of the 6GB unmapped **Quiver** reads with **CANU**, **MiniAsm** and I could not get more than 8Mb of contigs.  They seem to be very hard to assemble -- as expected.

I am still running **CANU** on the original reads with various params. We just discovered that we have ~10% chloroplast/mitochondrion contamination (despite the claim that the sample was pure gDNA). I will get rid of these reads from the input and try again.
